payments/models.py

```diff
@@ -154,3 +171,374 @@ def __str__(self):
+        ContentType,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="lease",
+    )
+    _lease_object_id = models.UUIDField(null=True, blank=True, verbose_name=_("lease"))
+
+    objects = OrderManager()
+
+    def __str__(self):
+        return f"{self.product} [{self.status}]"
+
+    @property
+    @rounded
+    def pretax_price(self):
+        return convert_aftertax_to_pretax(self.price, self.tax_percentage)
+
+    @property
+    @rounded
+    def total_price(self):
+        return sum([ol.price for ol in self.order_lines.all()], self.price)
+
+    @property
+    @rounded
+    def total_pretax_price(self):
+        return sum(
+            [ol.pretax_price for ol in self.order_lines.all()], self.pretax_price
+        )
+
+    @property
+    @rounded
+    def total_tax_percentage(self):
+        return ((self.total_price - self.total_pretax_price) / self.total_price) * 100
+
+    def _check_valid_products(self) -> None:
+        if self.product and not isinstance(
+            self.product, (BerthProduct, WinterStorageProduct)
+        ):
+            raise ValidationError(_("You cannot assign other types of products"))
+
+    def _check_valid_lease(self) -> None:
+        # Check that only Berth or Winter leases are passed (not other models)
+        if not isinstance(self.lease, (BerthLease, WinterStorageLease)):
+            raise ValidationError(_("You cannot assign other types of leases"))
+
+        # Check that the lease customer and the received customer are the same
+        if self.lease.customer != self.customer:
+            raise ValidationError(
+                _("The lease provided belongs to a different customer")
+            )
+
+    def _check_product_and_lease(self) -> None:
+        if isinstance(self.product, BerthProduct) and not isinstance(
+            self.lease, BerthLease
+        ):
+            raise ValidationError(
+                _("A BerthProduct must be associated with a BerthLease")
+            )
+        elif isinstance(self.product, WinterStorageProduct) and not isinstance(
+            self.lease, WinterStorageLease
+        ):
+            raise ValidationError(
+                _("A WinterStorageProduct must be associated with a WinterStorageLease")
+            )
+
+    def _check_product_or_price(self) -> None:
+        if all([not self.product, not self._product_object_id, not self.price]):
+            raise ValidationError(
+                _("Order must have either product object or price value")
+            )
+
+    def _check_product_and_price(self) -> None:
+        if (
+            self.price
+            and self.tax_percentage
+            and (self.product or self._product_object_id)
+        ):
+            raise ValidationError(
+                _("Cannot receive product and plain price/tax_percentage")
+            )
+
+    def _check_same_product(self, old_instance) -> None:
+        if old_instance.product != self.product:
```

Comment:
should this be `if old_instance.product and old_instance.product != self.product:`? or what did we agree on?

Reply:
Hmmm... :big-hm:

`old_instance` _should_ always be `not None`

```python
if self._state.adding:
    # Don't allow creating with product and price
    self._check_product_and_price()
else:
    old_instance = Order.objects.get(id=self.id)
    # If the product is being changed
    self._check_same_product(old_instance)
```

So `_check_same_product` (or `_check_same_lease`) are only called when the instance is updating.

Reply:
yeah, but the old instance might have NO product ☝️

Reply:
In that case, `if old_instance.product != self.product ---> None != BerthProduct()`

Reply:
```python
if all([old_instance.product, self.product, old_instance.product != self.product]):
    raise ValidationError()
```
like this? 🤔

Reply:
Sure 🤷 worth checking for everything.